Validation Documentation (#507)
jondavid-black/AaC: docs/docs/dev_guide/validation_plugins.md

---
layout: default
parent: Developer's Guide to AaC
title: Validation Plugins for Developers
nav_order: 7
has_children: false
---

# Validation Plugins for Developers

## Plugin Validations

### General Information surrounding Validation

Any references to code within this document regarding the `validation` plugin can be found here in the AaC project:
`AaC/python/src/validate/_validate.py`

This validator plugin can be separated into the below sections.

1. [Definition Validations](#definition-validations)
2. [Primitive Validations](#primitive-validations)

There is also a section listing out the `validation` plugin for the implementation of a validator plugin for 1st and/or 3rd party plugins

* [Implementation of a Validation Plugin for Definitions](#implementation-of-a-validation-plugin-for-definitions)

> It is also worthy to note that the way that the AaC DSL validates itself is through associating user-defined validation definitions with a corresponding python implementationn which is used to programmatically test that the validation's constraints are met. The validation process is entirely data-driven, so with this implementation it can validate the data in the definition via the very constraints.that are defined as part of the definition.

Reviewer: ```suggestion
It is also worthy to note that the way that the AaC DSL validates itself is through associating user-defined validation definitions with a corresponding python implementation which is used to programmatically test that the validation's constraints are met. The validation process is entirely data-driven, so with this implementation it can validate the data in the definition via the very constraints that are defined as part of the definition.
```

Author: Pushed through a commit that addresses this

  
     validator_results.append(_validate_primitive_types(definition, language_context)) 
     return validator_results 
```

With the above snippet this shows an example of how the validator is taking in a `Definition`, a list of other validator plugins, as well as, the `LanguageContext` of the definition that is being validated. The `Definition` validation is limited by *Structural Constraints*. With this in mind the`Definition` validator will take a `definition` that is being passed in and will decompose it into its `sub-definition`s within the primary schema. With this, this validator will take a definition and break it down to the structures that are present within. These `sub-structure`s are then validated by passing the sub-structure and corresponding  values to the appropriate validator which may produce finding(s) stored in a `ValidatorResult` object.

`ValidatorResult` object is then parsed through and returned to the user through some output of the latter results. This is also done for the below example in [Primitive Validations](#primitive-validations), but it is primarily targeted towards the `Primitive Types` within the `Definition`s.

### Primitive Validations

> Primitive Validations are passed into the below function:

Reviewer: Slightly more accurate description.
```suggestion
Primitive Validations are handled in the below function:
```

```python
def _validate_primitive_types(definition: Definition, language_context: LanguageContext) -> ValidatorResult: 
     """Validates the instances of AaC primitive types.""" 
     findings = ValidatorFindings() 
     definition_schema = get_definition_schema(definition, language_context) 
     if definition_schema: 
         findings.add_findings(_validate_fields(definition, definition_schema, definition.get_top_level_fields(), language_context)) 
  
     return ValidatorResult([definition], findings) 

```

The above method like the `_validate_definition` function outlined in the previous example takes in a `Definition` and takes in the `LanguageContext` of the defintion that is being validated. However, The primary goal of this snippet is to validate the *Primitive Types* of a definition schema and pass these results to the `ValidatorResult` object for further processing and eventual output to the user. This function is an example of a *Primitive Constraint* with how validations are made/applied.

> The results of this validation are also coinsiding with the results from the `_validate_definition` function above and the results are taking in together to the `ValidatorResult` object for further analysis and output for the user to discern.

Reviewer: Slight reword suggestion.
```suggestion
The results of this validation are consolidated with the results from the `_validate_definition` function above in the `ValidatorResult` object for further analysis and output for the user to discern.
```

As long as the arguments that are passed through the validator plugins and the returned types are similar then the plugins that are created/used can be used interchangeably and can be utilized based on the usecase.

An example of one of the validator plugins is the `bool_validator.py` plugin that can be found in the `AaC/python/src/aac/plugins/first_party/primitive_type_check/validators/`

```python
def get_validator() -> PrimitiveValidationContribution:
    """Return the Primitive Validator for 'boolean'."""
    return PrimitiveValidationContribution(BOOL_VALIDATION_NAME, PRIMITIVE_TYPE_BOOL, validate_bool)

def validate_bool(definition: Definition, value_to_validate: Any) -> Optional[ValidatorFinding]:
    """
    Returns a Validation finding if the type isn't valid, otherwise None.

    This function is intended to be used in the Validation apparatus, and for this result
    to be collated with other finding into a larger ValidatorResult.

    Args:
        definition (Definition): The definition that the value belongs to.
        value_to_validate (Any): The value to be tested.

    Returns:
        A ValidatorFinding if the value is not a boolean or None.
    """

```

> The return of this validator plugin is the `finding` from the `validate_bool` function. The arguments that are passed through is, the `Definition` and the type of the `value_to_validate` which is checking if the boolean that is being passed is valid. 

Reviewer: ```suggestion
The return of this validator plugin is the `finding` from the `validate_bool` function. The arguments that are passed through are, the `Definition` and the type of the `value_to_validate` which is checking if the boolean that is being passed is valid. 
```

Another validator is the `_validate_root_keys.py` plugin which checks and validates the root keys of the definition or schema. An example snippet of the arguments passed are as follows:

```python
def validate_root_keys(definition_under_test: Definition, target_schema_definition: Definition, language_context: LanguageContext, *validation_args) -> ValidatorResult:
    """
    Validates that the definition root key is defined by the root definition.

    Args:
        definition_under_test (Definition): The definition that's being validated.
        target_schema_definition (Definition): A definition with applicable validation.
        language_context (LanguageContext): The language context.

    Returns:
        A ValidatorResult containing any applicable error messages.
    """
```

Since this plugin is a part of the validator as a whole, it differs slightly from the `bool_validator.py` plugin that is a part of the `primitive_type_check` plugin. 

The arguments that are being passed through are similar, `Definition`, `LanguageContext`, and then any additional `validation_args`. As seen in the `_validate.py` validator the arguments and the return are quite similar.

> As mentioned before earlier the `ValidatorResult` will return the results of the validator from the functions that are being enacted, throughout the validation process, and compiles the results to a final message to notify the user of any `errors`, `warnings` or relevant informational messages regarding the validation.

Reviewer: ```suggestion
As mentioned before, the `ValidatorResult` will return the results of the validator from the functions that are being invoked, through the validation process, and compile the results into a final message that notifies the user of any `errors`, `warnings` or relevant informational messages regarding the validation.
```

### Best Practices for Validation Plugin Development

Best practices for the Validation Plugin Development is to do the following:

1. Keep the scope of the validator plugin small in scope; This allows for some flexibility and ease of testing the plugin.

> 2. Use `arguments` to validate; This allows for some flexibility for what can be validated.

Reviewer: ```suggestion
2. Use `arguments`; They allows for some flexibility for what can be validated.
```
